### PR TITLE
Add scikit-learn training for cmle notebook

### PR DESCRIPTION
**Pull request by nnegrey**

### CODE REVIEW

File: sklearn/notebooks/ml_engine_training.ipynb [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+    "  --scale-tier BASIC"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# [Optional] StackDriver Logging\n",
+    "You can view the logs for your training job:\n",
+    "1. Go to https://pantheon.corp.google.com \n",
```
Comment (sirtorry) [inline sklearn/notebooks/ml_engine_training.ipynb]: change to cloud console url before merge

File: sklearn/notebooks/ml_engine_training.ipynb [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "env: PROJECT_ID=true-ability-192918\n",
```
Comment (puneith) [inline sklearn/notebooks/ml_engine_training.ipynb]: remove true-ability* project from here

File: sklearn/notebooks/ml_engine_training.ipynb [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+    "\n",
+    "# TODO: REPLACE 'true-ability-192918' with your GCS BUCKET_ID\n",
+    "BUCKET_ID = 'true-ability-192918'\n",
+    "\n",
+    "\n",
+    "# ---------------------------------------\n",
+    "# 1. Add code to download the data from GCS (in this case, using the publicly hosted data).\n",
+    "# ML Engine will then be able to use the data when training your model.\n",
+    "# ---------------------------------------\n",
+    "census_data_filename = 'adult.data'\n",
```
Comment (puneith) [inline sklearn/notebooks/ml_engine_training.ipynb]: why are we downloading the file from GCS and then loading it in pandas?

File: sklearn/notebooks/ml_engine_training.ipynb (hunk 1 above)
Comment (nnegrey) [reply sklearn/notebooks/ml_engine_training.ipynb]: Right!

File: sklearn/notebooks/ml_engine_training.ipynb (hunk 3 above)
Comment (nnegrey) [reply sklearn/notebooks/ml_engine_training.ipynb]: On the CMLE machine.
1. Download the data locally for the machine
2. Use the data for the model.

File: sklearn/notebooks/census_training/train.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+pipeline = Pipeline([
+    ('union', preprocess),
+    ('classifier', classifier)
+])
+
+
+# ---------------------------------------
+# 2. Export and save the model to GCS
+# ---------------------------------------
+
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: # [START export-to-gcs]

File: sklearn/notebooks/census_training/train.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+joblib.dump(pipeline, model)
+
+# Upload the model to GCS
+model_path = os.path.join(
+    'gs://', 
+    BUCKET_ID, 
+    datetime.datetime.now().strftime('census_%Y%m%d_%H%M%S'),
+    model)
+
+subprocess.check_call(['gsutil', 'cp', model, model_path])
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: Add `# [END export-to-gcs]`

File: sklearn/notebooks/census_training/train.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
+# %load iris.py
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: Add `# [START setup]`

File: sklearn/notebooks/census_training/train.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+from sklearn.externals import joblib
+from sklearn.feature_selection import SelectKBest
+from sklearn.pipeline import FeatureUnion
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import LabelBinarizer
+
+
+# TODO: REPLACE 'true-ability-192918' with your GCS BUCKET_ID
+BUCKET_ID = 'true-ability-192918'
+
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: Add `# [END setup]`

File: sklearn/notebooks/census_training/train.py (hunk 6 above)
Comment (thedriftofwords) [reply sklearn/notebooks/census_training/train.py]: The "setup" region tag will include the imports and environment variable(s).

File: sklearn/notebooks/census_training/train.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+# ---------------------------------------
+# 1. Add code to download the data from GCS (in this case, using the publicly hosted data).
+# ML Engine will then be able to use the data when training your model.
+# ---------------------------------------
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: Add `# [START download-data]`

File: sklearn/notebooks/census_training/train.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+census_test_filename = 'adult.test'
+data_dir = 'gs://cloud-samples-data/ml-engine/sklearn/census_data'
+subprocess.check_call(['gsutil', 'cp',
+                       os.path.join(data_dir,census_data_filename),
+                       census_data_filename])
+subprocess.check_call(['gsutil', 'cp',
+                       os.path.join(data_dir, census_test_filename),
+                       census_test_filename])
+
+
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: Add `# [END download-data]`

File: sklearn/notebooks/census_training/train.py (hunk 8 above)
Comment (thedriftofwords) [reply sklearn/notebooks/census_training/train.py]: The "download-data" region tag encloses the "step 1" you defined in comments.

File: sklearn/notebooks/census_training/train.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+# ---------------------------------------
+# This is where your model code would go.
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: Is it plausible to put an example of "model code" in comments here? Or at least more of an explanation of exactly what you mean by "model code".

File: sklearn/notebooks/census_training/train.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+# ---------------------------------------
+
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: Add `# [START define-and-load-data]`

File: sklearn/notebooks/census_training/train.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+# Load the test census dataset
+with open('./adult.test', 'r') as test_data:
+    raw_testing_data = pd.read_csv(test_data, names=COLUMNS, skiprows=1)
+# Remove the column we are trying to predict ('income-level') from our features list
+# Convert the Dataframe to a lists of lists
+test_features = raw_testing_data.drop('income-level', axis=1).as_matrix().tolist()
+# Create our training labels list, convert the Dataframe to a lists of lists
+test_labels = (raw_testing_data['income-level'] == ' >50K.').as_matrix().tolist()
+
+
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: Add # [END define-and-load-data]

File: sklearn/notebooks/census_training/train.py [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+# Since the census data set has categorical features, we need to convert
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: Add `# [START categorical-feature-conversion]

Comment (thedriftofwords) [reply sklearn/notebooks/census_training/train.py]: (Before the start of your comment on that section)

File: sklearn/notebooks/census_training/train.py [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+        # Convert the categorical column to a numerical value
+        lbn = LabelBinarizer()
+        r = skb.transform(train_features)
+        lbn.fit(r)
+        # Create the pipeline to extract the categorical feature
+        categorical_pipelines.append(
+            ('categorical-{}'.format(i), Pipeline([
+                ('SKB-{}'.format(i), skb),
+                ('LBN-{}'.format(i), lbn)])))
+
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: Add # [END categorical-feature-conversion]

File: sklearn/notebooks/census_training/train.py [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+# Create pipeline to extract the numerical features
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: Before the start of the comment add `# [START create-pipeline]`

File: sklearn/notebooks/census_training/train.py [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+
+# Transform the features and fit them to the classifier
+classifier.fit(preprocess.transform(train_features), train_labels)
+
+# Create the overall model as a single pipeline
+pipeline = Pipeline([
+    ('union', preprocess),
+    ('classifier', classifier)
+])
+
```
Comment (thedriftofwords) [inline sklearn/notebooks/census_training/train.py]: Add `# [END create-pipeline]`

File: sklearn/notebooks/census_training/train.py (hunk 10 above)
Comment (nnegrey) [reply sklearn/notebooks/census_training/train.py]: The model code is everything below. 

I'll switch to: "This is where your model code would go. Below is an example model using the census dataset.

File: sklearn/notebooks/ml_engine_training.ipynb [hunk 17]
```diff
@@ -0,0 +1,507 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Part 0: Setup\n",
+    "* [Create a project on GCP](https://cloud.google.com/resource-manager/docs/creating-managing-projects)\n",
+    "* [Create a Google Cloud Storage Bucket](https://cloud.google.com/storage/docs/quickstart-console)\n",
+    "* [Enable Cloud Machine Learning Engine and Compute Engine APIs](https://console.cloud.google.com/flows/enableapi?apiid=ml.googleapis.com,compute_component&_ga=2.217405014.1312742076.1516128282-1417583630.1516128282)\n",
+    "* [Install Cloud SDK](https://cloud.google.com/sdk/downloads)\n",
+    "* [Install scikit-learn](http://scikit-learn.org/stable/install.html)\n",
+    "* [Install NumPy](https://docs.scipy.org/doc/numpy/user/install.html)\n",
```
[the remark this answers is not in the document]
Comment (nnegrey) [reply sklearn/notebooks/ml_engine_training.ipynb]: Done

File: sklearn/notebooks/ml_engine_training.ipynb [hunk 18]
```diff
@@ -0,0 +1,507 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Part 0: Setup\n",
+    "* [Create a project on GCP](https://cloud.google.com/resource-manager/docs/creating-managing-projects)\n",
+    "* [Create a Google Cloud Storage Bucket](https://cloud.google.com/storage/docs/quickstart-console)\n",
+    "* [Enable Cloud Machine Learning Engine and Compute Engine APIs](https://console.cloud.google.com/flows/enableapi?apiid=ml.googleapis.com,compute_component&_ga=2.217405014.1312742076.1516128282-1417583630.1516128282)\n",
+    "* [Install Cloud SDK](https://cloud.google.com/sdk/downloads)\n",
+    "* [Install scikit-learn](http://scikit-learn.org/stable/install.html)\n",
```
[the remark this answers is not in the document]
Comment (nnegrey) [reply sklearn/notebooks/ml_engine_training.ipynb]: I realized that they also could sort of run everything in the cloud and not even run the files locally.
So I could drop it.

File: sklearn/notebooks/ml_engine_training.ipynb [hunk 19]
```diff
@@ -0,0 +1,507 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Part 1: Create your python model file\n",
+    "\n",
+    "First, we'll create the python model file (provided below) that we'll upload to ML Engine. This is similar to your normal process for creating a scikit-learn model. However, there are two key differences:\n",
+    "1. Downloading the data from GCS at the start of your file, so that ML Engine can access the data.\n",
+    "1. Exporting/saving the model to GCS at the end of your file, so that you can use it for predictions.\n",
+    "\n",
+    "The model in this file loads the the data into a pandas DataFrame that can be used by scikit-learn. Then the model is fit against the training data. Lastly, sklearn's built in version of joblib is used to save the model to a file that can be uploaded to [ML Engine's prediction service](https://cloud.google.com/ml-engine/docs/scikit/getting-predictions#deploy_models_and_versions).\n",
```
[the remark this answers is not in the document]
Comment (nnegrey) [reply sklearn/notebooks/ml_engine_training.ipynb]: fixed

File: sklearn/notebooks/census_training/train.py [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+    raw_training_data = pd.read_csv(train_data, header=None, names=COLUMNS)
+
+# Remove the column we are trying to predict ('income-level') from our features list
+# Convert the Dataframe to a lists of lists
+train_features = raw_training_data.drop('income-level', axis=1).as_matrix().tolist()
+# Create our training labels list, convert the Dataframe to a lists of lists
+train_labels = (raw_training_data['income-level'] == ' >50K').as_matrix().tolist()
+
+
+# Load the test census dataset
```
[the remark this answers is not in the document]
Comment (nnegrey) [reply sklearn/notebooks/census_training/train.py]: Fixed

File: sklearn/notebooks/ml_engine_training.ipynb (hunk 18 above)
[the remark this answers is not in the document]
Comment (sirtorry) [reply sklearn/notebooks/ml_engine_training.ipynb]: I think we should let them know you need those libraries if you choose to explore models locally (which we recommend). but it's not necessary for the task so if you're not looking to play around locally, there's no need.

[the remark this answers is not in the document]
Comment (nnegrey) [reply sklearn/notebooks/ml_engine_training.ipynb]: How about: install scikit-learn [Optional: used if running locally]

[the remark this answers is not in the document]
Comment (sirtorry) [reply sklearn/notebooks/ml_engine_training.ipynb]: yup, that's what i'm doing for xgboost notebook

File: sklearn/notebooks/ml_engine_training.ipynb (hunk 2 above)
Comment (sirtorry) [reply sklearn/notebooks/ml_engine_training.ipynb]: @nnegrey bump

File: sklearn/notebooks/ml_engine_training.ipynb [hunk 21]
```diff
@@ -0,0 +1,489 @@
...
+    "# ML Engine will then be able to use the data when training your model.\n",
+    "# ---------------------------------------\n",
+    "# [START download-data]\n",
+    "census_data_filename = 'adult.data'\n",
+    "\n",
+    "# Public bucket holding the census data\n",
+    "bucket = storage.Client().bucket('cloud-samples-data')\n",
+    "\n",
+    "# Path to the data inside the public bucket\n",
+    "data_dir = 'ml-engine/sklearn/census_data/'\n",
```
Comment (sirtorry) [inline sklearn/notebooks/ml_engine_training.ipynb]: switch to gs://cloud-samples-data/ml-engine/census/data/adult.data.csv ?

File: sklearn/notebooks/ml_engine_training.ipynb [hunk 22]
```diff
@@ -0,0 +1,489 @@
...
+    "* [Install Cloud SDK](https://cloud.google.com/sdk/downloads)\n",
+    "* [Install scikit-learn](http://scikit-learn.org/stable/install.html)\n",
+    "* [Install pandas](https://pandas.pydata.org/pandas-docs/stable/install.html)\n",
+    "\n",
+    "These variables will be needed for the following steps.\n",
+    "* `TRAINER_PACKAGE_PATH <./census_training>` - A packaged training application that will be staged in a Google Cloud Storage location. The model file created below is placed inside this package path.\n",
+    "* `MAIN_TRAINER_MODULE <census_training.train>` - Tells ML Engine which file to execute. This is formatted as follows <folder_name.python_file_name>\n",
+    "* `JOB_DIR <gs://$BUCKET_ID/scikit_learn_job_dir>` - The path to a Google Cloud Storage location to use for job output.\n",
+    "* `RUNTIME_VERSION <1.8>` - The version of Cloud ML Engine to use for the job. If you don't specify a runtime version, the training service uses the default Cloud ML Engine runtime version 1.0. [See the list of runtime versions for more information](https://cloud.google.com/ml-engine/docs/tensorflow/runtime-version-list).\n",
```
Comment (sirtorry) [inline sklearn/notebooks/ml_engine_training.ipynb]: 1.9 to not confuse people.

File: sklearn/notebooks/ml_engine_training.ipynb [hunk 23]
```diff
@@ -0,0 +1,489 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# [Optional] Verify Model File in GCS\n",
+    "View the contents of the destination model folder to verify that model file has indeed been uploaded to GCS.\n",
+    "\n",
+    "Note: This can take a few minutes."
```
Comment (sirtorry) [inline sklearn/notebooks/ml_engine_training.ipynb]: Note: The model can take a few minutes to train and show up in GCS.

File: sklearn/notebooks/ml_engine_training.ipynb (hunk 21 above)
Comment (nnegrey) [reply sklearn/notebooks/ml_engine_training.ipynb]: Done.

File: sklearn/notebooks/ml_engine_training.ipynb (hunk 22 above)
Comment (nnegrey) [reply sklearn/notebooks/ml_engine_training.ipynb]: Fixed

File: sklearn/notebooks/ml_engine_training.ipynb (hunk 23 above)
Comment (nnegrey) [reply sklearn/notebooks/ml_engine_training.ipynb]: Updated

**Pull request merged**